### PR TITLE
No inherited threadlocal

### PR DESCRIPTION
**Pull request by elmot**
This change is [Reviewable](https://reviewable.io/reviews/vaadin/framework/96)

### CODE REVIEW

File: server/src/main/java/com/vaadin/server/VaadinService.java [hunk 1]
```diff
@@ -1860,17 +1859,13 @@ public void runPendingAccessTasks(VaadinSession session) {
...
+        Map<Class<?>, CurrentInstance> oldInstances = CurrentInstance
+                .setCurrent(session);
         try {
             while ((pendingAccess = session.getPendingAccessQueue()
                     .poll()) != null) {
                 if (!pendingAccess.isCancelled()) {
-                    CurrentInstance.clearAll();
-                    CurrentInstance.restoreInstances(
-                            pendingAccess.getCurrentInstances());
                     CurrentInstance.setCurrent(session);
```
Comment (elmot) [inline server/src/main/java/com/vaadin/server/VaadinService.java]: This line was removed from the original change, but it seems to me it's better to have it here

**Comment by elmot** [comment]
Fixes vaadin/framework8-issues#514

**Comment by Legioth** [comment]
Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/src/main/java/com/vaadin/server/VaadinService.java, line 1869 at r1](https://reviewable.io:443/reviews/vaadin/framework/96#-KYJpNrTxaLFdLFMROh_-r1-1869:-KYNDGjk3R5uVNK-WHA1:b-sd76tg) ([raw file](https://github.com/vaadin/framework/blob/c2208f14fb99ebe58d83003b86e862aba730769b/server/src/main/java/com/vaadin/server/VaadinService.java#L1869)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

This line was removed from the original change, but it seems to me it's better to have it here
</blockquote></details>

Better in what way?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by elmot** [comment]
Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/src/main/java/com/vaadin/server/VaadinService.java, line 1869 at r1](https://reviewable.io:443/reviews/vaadin/framework/96#-KYJpNrTxaLFdLFMROh_-r1-1869:-KYOTXjT5Z1ywNHYLkHv:bpjl6r8) ([raw file](https://github.com/vaadin/framework/blob/c2208f14fb99ebe58d83003b86e862aba730769b/server/src/main/java/com/vaadin/server/VaadinService.java#L1869)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Better in what way?
</blockquote></details>

Well after one more review of all the use cases, I'd say you are right. No sense, removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by hesara** [comment]
Reviewed 10 of 13 files at r1.
Review status: 10 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[server/src/main/java/com/vaadin/server/VaadinService.java, line 1871 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXSVGHCbWMG3-3Q4V1:-KYXSVGI815Zb-TiyQUC:b-n8fpxq) ([raw file](https://github.com/vaadin/framework/blob/5b0919ce9e3f8899fb5b0c5331afd9ffa2a56c0c/server/src/main/java/com/vaadin/server/VaadinService.java#L1871)):*
> ```Java
>                     .poll()) != null) {
>                 if (!pendingAccess.isCancelled()) {
>                     CurrentInstance.clearAll();
> ```

If I understand this correctly, removing these lines means that if one VaadinSession.access() modifies thread locals, those modified ones might or might not be effective in the next one depending on whether it happened to be in the queue at the same time. Is that intentional?

---

*[server/src/main/java/com/vaadin/server/VaadinServlet.java, line 305 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXNZJ7CFjOvuaTqRd2:-KYXNZJ8Q-AoPYxxxaSs:b-5zpq4b) ([raw file](https://github.com/vaadin/framework/blob/5b0919ce9e3f8899fb5b0c5331afd9ffa2a56c0c/server/src/main/java/com/vaadin/server/VaadinServlet.java#L305)):*
> ```Java
>      * servlet is not automatically defined.
>      * <p>
>      * The current servlet is derived from the current service using
> ```

Is this information no longer relevant?

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 47 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXP_NcWqqZNLudpEcv:-KYXP_Ndg6jtaY5ZEH04:bt5653q) ([raw file](https://github.com/vaadin/framework/blob/ee446c0eaff7f896828d03c6beaae26645df3a36/server/src/main/java/com/vaadin/util/CurrentInstance.java#L47)):*
> ```Java
>  * </p>
>  * <p>
>  * {@link UI}, {@link VaadinService}, {@link VaadinSession}.
> ```

Should this be a comma as the two lists are no longer that separate?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by hesara** [comment]
Reviewed 2 of 13 files at r1.
Review status: 12 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 260 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXwo2FvxbKKn-m2f0N:-KYXwo2FvxbKKn-m2f0O:bnjw4l1) ([raw file](https://github.com/vaadin/framework/blob/ee446c0eaff7f896828d03c6beaae26645df3a36/server/src/main/java/com/vaadin/util/CurrentInstance.java#L260)):*
> ```Java
>     public static Map<Class<?>, CurrentInstance> setCurrent(UI ui) {
>         Map<Class<?>, CurrentInstance> old = setCurrent(ui.getSession());
>         old.put(UI.class, set(UI.class, ui));
> ```

How getting the old content and how put works etc. is quite unclear to me here and below while just reading these bits of code, but if we assume it works correctly in the old version, this looks ok to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by hesara** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by Legioth** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[server/src/main/java/com/vaadin/server/VaadinService.java, line 1871 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXSVGHCbWMG3-3Q4V1:-KYYCbwJkQ8MIfjjge6K:b-rrhw7l) ([raw file](https://github.com/vaadin/framework/blob/5b0919ce9e3f8899fb5b0c5331afd9ffa2a56c0c/server/src/main/java/com/vaadin/server/VaadinService.java#L1871)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

If I understand this correctly, removing these lines means that if one VaadinSession.access() modifies thread locals, those modified ones might or might not be effective in the next one depending on whether it happened to be in the queue at the same time. Is that intentional?
</blockquote></details>

The assumption is still that the task wouldn't make any changes. Otherwise the clearing would be done after instead of before calling the task. 

---

*[server/src/main/java/com/vaadin/server/VaadinServlet.java, line 305 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXNZJ7CFjOvuaTqRd2:-KYYDyCfXUF2xQ9FTSHP:blb28sg) ([raw file](https://github.com/vaadin/framework/blob/5b0919ce9e3f8899fb5b0c5331afd9ffa2a56c0c/server/src/main/java/com/vaadin/server/VaadinServlet.java#L305)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Is this information no longer relevant?
</blockquote></details>

I agree. Why is this removed? It hasn't been removed by any of the the original Hummingbird patches.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 47 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXP_NcWqqZNLudpEcv:-KYYETRol3sssT8KAU54:b-xi4pps) ([raw file](https://github.com/vaadin/framework/blob/ee446c0eaff7f896828d03c6beaae26645df3a36/server/src/main/java/com/vaadin/util/CurrentInstance.java#L47)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Should this be a comma as the two lists are no longer that separate?
</blockquote></details>

Yes, that's how it's done in the original patches in Hummingbird.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by Legioth** [comment]
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/src/main/java/com/vaadin/server/VaadinServlet.java, line 306 at r3](https://reviewable.io:443/reviews/vaadin/framework/96#-KYZSdWquScEFADHAuKJ:-KYZSdWquScEFADHAuKK:bu97qag) ([raw file](https://github.com/vaadin/framework/blob/bd2c99f7798e6c4348f3f074174ab5945050254e/server/src/main/java/com/vaadin/server/VaadinServlet.java#L306)):*
> ```Java
>      * <p>
>      * The current servlet is derived from the current service using
>      *
> ```

Misplaced empty line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by elmot** [comment]
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/src/main/java/com/vaadin/server/VaadinService.java, line 1871 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXSVGHCbWMG3-3Q4V1:-KYZOXjD5yWmWUT6nenw:b-oaxwmm) ([raw file](https://github.com/vaadin/framework/blob/5b0919ce9e3f8899fb5b0c5331afd9ffa2a56c0c/server/src/main/java/com/vaadin/server/VaadinService.java#L1871)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

The assumption is still that the task wouldn't make any changes. Otherwise the clearing would be done after instead of before calling the task. 
</blockquote></details>

What for the task should modify any threadlocals? Isn't that bad design? Do we really need to protect threadlocals here while developers who use threadlocals can do much more dangerous things?

---

*[server/src/main/java/com/vaadin/server/VaadinServlet.java, line 305 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXNZJ7CFjOvuaTqRd2:-KYZN6LVVJJHQft7C3Ft:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5b0919ce9e3f8899fb5b0c5331afd9ffa2a56c0c/server/src/main/java/com/vaadin/server/VaadinServlet.java#L305)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

I agree. Why is this removed? It hasn't been removed by any of the the original Hummingbird patches.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/server/VaadinServlet.java, line 306 at r3](https://reviewable.io:443/reviews/vaadin/framework/96#-KYZSdWquScEFADHAuKJ:-KYZW44h5WNxJu6yH6vH:buketii) ([raw file](https://github.com/vaadin/framework/blob/bd2c99f7798e6c4348f3f074174ab5945050254e/server/src/main/java/com/vaadin/server/VaadinServlet.java#L306)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Misplaced empty line
</blockquote></details>

Indeed

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 47 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXP_NcWqqZNLudpEcv:-KYZNFX1lOz81IoPFj9o:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/ee446c0eaff7f896828d03c6beaae26645df3a36/server/src/main/java/com/vaadin/util/CurrentInstance.java#L47)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Yes, that's how it's done in the original patches in Hummingbird.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by hesara** [comment]
Review status: 11 of 13 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/src/main/java/com/vaadin/server/VaadinService.java, line 1871 at r2](https://reviewable.io:443/reviews/vaadin/framework/96#-KYXSVGHCbWMG3-3Q4V1:-KYmSBOwqEZOHQhLrVro:b-s8kzr7) ([raw file](https://github.com/vaadin/framework/blob/5b0919ce9e3f8899fb5b0c5331afd9ffa2a56c0c/server/src/main/java/com/vaadin/server/VaadinService.java#L1871)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

What for the task should modify any threadlocals? Isn't that bad design? Do we really need to protect threadlocals here while developers who use threadlocals can do much more dangerous things?
</blockquote></details>

It is bad application design if a task modifies threadlocals, but I have the impression I just saw a ticket somewhere that seems to originate either from such a situation or from a bug in the restore mechanism. Anyway, maybe we can leave this outside the scope of this patch if the finally block below cleans things up correctly after the whole queue has been processed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by hesara** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96)*

**Comment by hesara** [comment]
:lgtm: - any old issues with CurrentInstance can be dealt with separately

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/96#-:-KYrIrmIgyySsaiwLMOr:b-bzeyva)*

**Pull request merged**